io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3EMImpl.java

```diff
@@ -0,0 +1,270 @@
+				this.channel(IoShellyPlus3EM.ChannelId.RELAY) //
+		};
+
+		ElectricityMeter.calculateSumActivePowerFromPhases(this);
+	}
+
+	@Activate
+	protected void activate(ComponentContext context, Config config) {
+	    super.activate(context, config.id(), config.alias(), config.enabled());
+	    this.meterType = config.type();
+	    this.baseUrl = "http://" + config.ip();
+
+	    if (this.isEnabled()) {
+	        this.httpBridge.subscribeJsonEveryCycle(this.baseUrl + "/status", this::processHttpResult);
+	    }
+	}
+
+	@Override
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+	}
+
+	@Override
+	public BooleanWriteChannel[] digitalOutputChannels() {
+		return this.digitalOutputChannels;
+	}
+
+	@Override
+	public String debugLog() {
+	    var b = new StringBuilder();
+	    var valueOpt = this.getRelayChannel().value().asOptional();
+	    if (valueOpt.isPresent()) {
+	        b.append(valueOpt.get() ? "ON" : "OFF");
+	    } else {
+	        b.append("Unknown");
+	    }
+	    b.append("|");
+	    b.append(this.getActivePowerChannel().value().asString());
+
+	    return b.toString();
+	}
+
+
+	@Override
+	public void handleEvent(Event event) {
+		if (!this.isEnabled()) {
+			return;
+		}
+
+		switch (event.getTopic()) {
+		case EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE -> this.calculateEnergy();
+		case EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE -> {
+			this.executeWrite(this.getRelayChannel(), 0);
+		}
+		}
+	}
+	
+	
+	private void processHttpResult(JsonElement result, Throwable error) {
+	    
+	    this._setSlaveCommunicationFailed(result == null);
+	    if (error != null) {
+	        this._setRelay(null);
+	        this._setActivePower(null);
+	        this._setActiveProductionEnergy(null);
+	        this.logDebug(this.log, error.getMessage());
+	        return;
+	    }
+
+	    try {
+	        JsonObject jsonResponse = result.getAsJsonObject();
+	        
+
+	        JsonArray relays = jsonResponse.getAsJsonArray("relays");
+	        if (relays != null && relays.size() > 0) {
+	            JsonObject relay = relays.get(0).getAsJsonObject();
+	            boolean isOn = relay.get("ison").getAsBoolean();
+	            this.getRelayChannel().setNextWriteValue(isOn);
+	            this._setRelay(isOn);
+
+	        }
+	        
+	        
+	        JsonObject updateObject = jsonResponse.getAsJsonObject("update");
+	        if (updateObject != null) {
+	            boolean hasUpdate = updateObject.get("has_update").getAsBoolean();
+				this.channel(IoShellyPlus3EM.ChannelId.HAS_UPDATE).setNextValue(hasUpdate);
+	        }
+
+	        JsonArray emeters = jsonResponse.getAsJsonArray("emeters");
+	        for (int i = 0; i < emeters.size(); i++) {
+	            JsonObject emeter = emeters.get(i).getAsJsonObject();
+	            Float power = emeter.get("power").getAsFloat();
+	            Float voltage = emeter.get("voltage").getAsFloat();
+	            Float current = emeter.get("current").getAsFloat();
```

Comment:
please use primitives there is a difference between the type `float` and `Float` and its better to use primitives most of the times. You can read more about it [here](https://www.baeldung.com/java-primitives-vs-objects)

Reply:
I also always recommend to use the appropriate JsonUtils methods, because they handle Exceptions better.